File: src/base_control/base_control/md2_transport.py
"""Modbus Transport — Multi-drive 2.0 over RS-485.

負責 Modbus RTU 封包組裝、CRC、序列埠收發。

不負責：單位換算、運動學、驅動器狀態機（由上層 driver_interface 處理）。

協議參數於 2026-08-07 實機驗證，詳見 docs/implementation/SUB-001-base-control-plan.md。
"""

from __future__ import annotations

import struct
import time
from dataclasses import dataclass

import serial

# ── Modbus function codes ────────────────────────────────────────────────────
FC_READ_HOLDING = 0x03
FC_WRITE_SINGLE = 0x06
FC_READ_WRITE_MULTIPLE = 0x17

# ── Multi-drive 2.0 群組定址 ─────────────────────────────────────────────────
# 位址上位元組 [15:12]=0xF、[11:8]=Index；下位元組 bit n = Driver ID (n+1) 被選中。
MD2_GROUP_ID = 0x65
_DRIVER_BITMASK = 0x03          # ID1 + ID2

READ_INDEX_BASE = 0
READ_ITEM_COUNT = 7             # Read index 0~6
WRITE_INDEX_BASE = 8
WRITE_ITEM_COUNT = 2            # Write index 8~9

R_ADDR = 0xF000 | (READ_INDEX_BASE << 8) | _DRIVER_BITMASK    # 0xF003
W_ADDR = 0xF000 | (WRITE_INDEX_BASE << 8) | _DRIVER_BITMASK   # 0xF803

NUM_DRIVERS = 2
_ITEMS_PER_DRIVER = READ_ITEM_COUNT + 1     # +1: Error_Check
R_CNT = _ITEMS_PER_DRIVER * NUM_DRIVERS     # 16 words
W_CNT = WRITE_ITEM_COUNT * NUM_DRIVERS      # 4 words

# Read Data Mapping（09-26 = 0）
IDX_STATUS = 0
IDX_ALARM = 1
IDX_RPM = 2
IDX_POS_TURNS = 5
IDX_POS_PULSE = 6


class Md2Error(Exception):
    """Multi-drive 2.0 通訊或解析錯誤。"""


@dataclass(frozen=True, slots=True)
class DriverRaw:
    """單一驅動器之原始回授（未套用方向修正與單位換算）。"""

    status: int
    alarm: int
    rpm: int            # signed, 馬達端 RPM
    pos_turns: int      # signed, 圈數
    pos_pulse: int      # 0 ~ (pulse_per_rev - 1)


@dataclass(frozen=True, slots=True)
class Md2Feedback:
    """一次 FC17h 交易之雙驅動器回授。"""

    right: DriverRaw
    left: DriverRaw
    comm_s: float


def crc16(data: bytes) -> int:
    crc = 0xFFFF
    for byte in data:
        crc ^= byte
        for _ in range(8):
            crc = (crc >> 1) ^ 0xA001 if crc & 1 else crc >> 1
    return crc


def _with_crc(frame: bytes) -> bytes:
    crc = crc16(frame)
    return frame + bytes([crc & 0xFF, (crc >> 8) & 0xFF])


def _crc_ok(frame: bytes) -> bool:
    return len(frame) >= 3 and crc16(frame[:-2]) == (frame[-2] | (frame[-1] << 8))


def to_s16(value: int) -> int:
    value &= 0xFFFF
    return value - 0x10000 if value >= 0x8000 else value


class Md2Transport:
    """RS-485 Multi-drive 2.0 傳輸層。"""

    def __init__(
        self,
        port: str,
        baud: int = 230400,
        timeout_s: float = 0.1,
        bytesize: int = 8,
        parity: str = 'N',
        stopbits: int = 1,
    ) -> None:
        self._serial = serial.Serial(
            port=port,
            baudrate=baud,
            bytesize=bytesize,
            parity=parity,
            stopbits=stopbits,
            timeout=timeout_s,
        )

    def close(self) -> None:
        if self._serial.is_open:
            self._serial.close()

    def drain(self, settle_s: float = 0.05) -> None:
        """清空收發緩衝並等待匯流排靜默。

        交易途中被中斷時，未讀完之回應會殘留於緩衝區，
        使後續請求讀到前一筆回應而 CRC 失敗。關閉流程前須先排空。
        """
        time.sleep(settle_s)
        self._serial.reset_input_buffer()
        self._serial.reset_output_buffer()

    def __enter__(self) -> 'Md2Transport':
        return self

    def __exit__(self, *_) -> None:
        self.close()

    # ── 低階收發 ─────────────────────────────────────────────────────────────

    def _transact(self, request: bytes, expect_len: int) -> bytes:
        self._serial.reset_input_buffer()
        self._serial.write(request)
        self._serial.flush()
        response = self._serial.read(expect_len)
        if len(response) < expect_len:
            raise Md2Error(
                f'回應長度不足：期望 {expect_len}，收到 {len(response)}'
                f'（{response.hex()}）'
            )
        if response[1] & 0x80:
            raise Md2Error(f'驅動器例外碼 0x{response[2]:02X}')
        if not _crc_ok(response):
            raise Md2Error(f'CRC 錯誤：{response.hex()}')
        return response

    # ── FC03 / FC06：個別驅動器 ──────────────────────────────────────────────

    def read_register(self, driver_id: int, address: int) -> int:
        """讀取單一保持暫存器。"""
        request = _with_crc(
            bytes([driver_id, FC_READ_HOLDING]) + struct.pack('>HH', address, 1)
        )
        response = self._transact(request, 7)
        return struct.unpack('>H', response[3:5])[0]

    def write_register(self, driver_id: int, address: int, value: int) -> None:
        """寫入單一保持暫存器。"""
        request = _with_crc(
            bytes([driver_id, FC_WRITE_SINGLE])
            + struct.pack('>HH', address, value & 0xFFFF)
        )
        self._transact(request, 8)

    # ── FC17h：雙驅動器讀寫合一 ──────────────────────────────────────────────

    def read_write(
        self,
        right_cmd: int,
        right_rpm: int,
        left_cmd: int,
        left_rpm: int,
    ) -> Md2Feedback:
        """單一封包同時下達雙輪命令並讀回回授。

        Write data 為 driver-major：[ID1_cmd][ID1_rpm][ID2_cmd][ID2_rpm]。
        """
        header = bytes([MD2_GROUP_ID, FC_READ_WRITE_MULTIPLE]) + struct.pack(
            '>HHHHB', R_ADDR, R_CNT, W_ADDR, W_CNT, W_CNT * 2
        )
        body = struct.pack(
            '>HHHH',
            right_cmd & 0xFFFF,
            right_rpm & 0xFFFF,
            left_cmd & 0xFFFF,
            left_rpm & 0xFFFF,
        )
        expect_len = 3 + R_CNT * 2 + 2

        start = time.monotonic()
        response = self._transact(_with_crc(header + body), expect_len)
        comm_s = time.monotonic() - start

        if response[2] != R_CNT * 2:
            raise Md2Error(
                f'byte_cnt 不符：期望 {R_CNT * 2}，收到 {response[2]}'
            )

        words = [
            struct.unpack('>H', response[3 + i * 2:5 + i * 2])[0]
            for i in range(R_CNT)
        ]
        return Md2Feedback(
            right=self._extract(words, 0),
            left=self._extract(words, 1),
            comm_s=comm_s,
        )

    @staticmethod
    def _extract(words: list[int], driver_slot: int) -> DriverRaw:
        base = driver_slot * _ITEMS_PER_DRIVER
        return DriverRaw(
            status=words[base + IDX_STATUS],
            alarm=words[base + IDX_ALARM],
            rpm=to_s16(words[base + IDX_RPM]),
            pos_turns=to_s16(words[base + IDX_POS_TURNS]),
            pos_pulse=words[base + IDX_POS_PULSE],
        )

File: src/base_control/src/md2_transport.cpp
#include "base_control/md2_transport.hpp"

#include <modbus/modbus.h>

#include <array>
#include <cerrno>
#include <chrono>
#include <cstring>
#include <string>

namespace base_control
{

namespace
{

std::string modbus_error(const std::string & context)
{
  return context + "：" + modbus_strerror(errno);
}

int16_t to_s16(uint16_t value)
{
  return static_cast<int16_t>(value);
}

}  // namespace

Md2Transport::~Md2Transport()
{
  close();
}

void Md2Transport::open(
  const std::string & port, int baud, char parity,
  int data_bits, int stop_bits, double response_timeout_s)
{
  if (ctx_ != nullptr) {
    throw Md2Error("序列埠已開啟，請先 close()");
  }

  ctx_ = modbus_new_rtu(port.c_str(), baud, parity, data_bits, stop_bits);
  if (ctx_ == nullptr) {
    throw Md2Error(modbus_error("建立 Modbus RTU context 失敗"));
  }

  if (modbus_connect(ctx_) == -1) {
    const std::string message = modbus_error("開啟序列埠 " + port + " 失敗");
    modbus_free(ctx_);
    ctx_ = nullptr;
    throw Md2Error(message);
  }

  const auto seconds = static_cast<uint32_t>(response_timeout_s);
  const auto micros =
    static_cast<uint32_t>((response_timeout_s - static_cast<double>(seconds)) * 1e6);
  modbus_set_response_timeout(ctx_, seconds, micros);
}

void Md2Transport::close()
{
  if (ctx_ == nullptr) {
    return;
  }
  modbus_close(ctx_);
  modbus_free(ctx_);
  ctx_ = nullptr;
}

void Md2Transport::drain()
{
  if (ctx_ != nullptr) {
    modbus_flush(ctx_);
  }
}

void Md2Transport::ensure_open() const
{
  if (ctx_ == nullptr) {
    throw Md2Error("序列埠尚未開啟");
  }
}

uint16_t Md2Transport::read_register(int driver_id, int address)
{
  ensure_open();
  if (modbus_set_slave(ctx_, driver_id) == -1) {
    throw Md2Error(modbus_error("設定 Driver ID " + std::to_string(driver_id) + " 失敗"));
  }

  uint16_t value = 0;
  if (modbus_read_registers(ctx_, address, 1, &value) == -1) {
    throw Md2Error(
      modbus_error(
        "讀取 Driver " + std::to_string(driver_id) +
        " 暫存器 0x" + std::to_string(address) + " 失敗"));
  }
  return value;
}

void Md2Transport::write_register(int driver_id, int address, uint16_t value)
{
  ensure_open();
  if (modbus_set_slave(ctx_, driver_id) == -1) {
    throw Md2Error(modbus_error("設定 Driver ID " + std::to_string(driver_id) + " 失敗"));
  }

  if (modbus_write_register(ctx_, address, value) == -1) {
    throw Md2Error(
      modbus_error(
        "寫入 Driver " + std::to_string(driver_id) +
        " 暫存器 0x" + std::to_string(address) + " 失敗"));
  }
}

Md2Feedback Md2Transport::read_write(
  uint16_t right_cmd, int16_t right_rpm,
  uint16_t left_cmd, int16_t left_rpm)
{
  ensure_open();
  if (modbus_set_slave(ctx_, kMd2GroupId) == -1) {
    throw Md2Error(modbus_error("設定群組位址失敗"));
  }

  // driver-major：[ID1_cmd][ID1_rpm][ID2_cmd][ID2_rpm]
  const std::array<uint16_t, kWriteCount> write_data{
    right_cmd,
    static_cast<uint16_t>(right_rpm),
    left_cmd,
    static_cast<uint16_t>(left_rpm),
  };
  std::array<uint16_t, kReadCount> read_data{};

  const auto start = std::chrono::steady_clock::now();
  const int rc = modbus_write_and_read_registers(
    ctx_,
    kWriteAddr, kWriteCount, write_data.data(),
    kReadAddr, kReadCount, read_data.data());
  const auto elapsed = std::chrono::steady_clock::now() - start;

  if (rc == -1) {
    throw Md2Error(modbus_error("FC17h 交易失敗"));
  }

  Md2Feedback feedback;
  feedback.comm_s = std::chrono::duration<double>(elapsed).count();

  const auto extract = [&read_data](int slot) {
      const int base = slot * kItemsPerDriver;
      DriverRaw raw;
      raw.status = read_data[base + kIdxStatus];
      raw.alarm = read_data[base + kIdxAlarm];
      raw.rpm = to_s16(read_data[base + kIdxRpm]);
      raw.pos_turns = to_s16(read_data[base + kIdxPosTurns]);
      raw.pos_pulse = read_data[base + kIdxPosPulse];
      return raw;
    };

  feedback.right = extract(0);
  feedback.left = extract(1);
  return feedback;
}

}  // namespace base_control
